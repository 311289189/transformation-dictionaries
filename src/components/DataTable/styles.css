* {
    transition: all 200ms ease-in-out;
}

.dataTable {
    padding: 1em;
    border: 1px solid #f0ebeb;
    border-radius: 5px;
    width: 500px;
}

.dataTable input {
    border-radius: 4px;
    font-size: 16px;
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    outline: none;
    line-height: 25px;
    border: none;
    background-color: rgb(240, 240, 240);
    width: 90%;
    padding: 0 10px;
}

.dataTable input:focus {
    background-color: rgba(240, 240, 240, 0.5);
}
