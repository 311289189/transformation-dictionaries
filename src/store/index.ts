import { action, observable } from 'mobx'

export type DictionaryLookup = Record<string, string>

export interface IStore {
    availableDictionaries: DictionaryLookup
}

class Store implements IStore {
    @observable availableDictionaries = {}

    @action.bound
    public addDictionaryItem(newItem: DictionaryLookup) {
        this.availableDictionaries = {
            ...this.availableDictionaries,
            ...newItem
        }
    }

    @action.bound
    public removeDictionaryItem(itemKey: string) {
        delete this.availableDictionaries[itemKey]
    }
}

export default new Store()
