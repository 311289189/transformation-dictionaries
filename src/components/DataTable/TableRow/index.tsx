import * as React from 'react'
import styles from './styles.css'
import { observer } from 'mobx-react-lite'
import { inject } from 'mobx-react'
import { IStore, Validation } from '../../../store'
import { DictionaryMapping } from '../index'
import { FormEvent } from 'react'

const classnames = require('classnames')

const checkmark = require('./checkmark.png')
const cross = require('./cross.png')

interface Props {
    store: IStore
    editable: boolean
    dictionaryItem: DictionaryMapping
    index: number
}

const TableRow = ({ store, dictionaryItem, editable, index }: Props) => {
    const isFirstRow = index === 0

    const [cacheItem, setCacheItem] = React.useState(dictionaryItem)
    const [inputs, setInput] = React.useState(dictionaryItem)

    const fromInput = React.useRef<HTMLInputElement>(null)
    const toInput = React.useRef<HTMLInputElement>(null)

    const runLocalInputValidations = () => {
        fromInput.current && fromInput.current.reportValidity()
        toInput.current && toInput.current.reportValidity()
    }

    const clearLocalInputValidations = () => {
        fromInput.current && fromInput.current.setCustomValidity('')
        toInput.current && toInput.current.setCustomValidity('')
    }

    const clearLocalInputFields = () => setInput({ from: '', to: '' })

    const resetLocalInputFrom = () => {
        store.clearValidation()
        !isFirstRow && setInput({ from: cacheItem.from, to: inputs.to })
    }

    const resetLocalInputTo = () => {
        store.clearValidation()
        !isFirstRow && setInput({ from: inputs.from, to: cacheItem.to })
    }

    React.useEffect(() => {
        if (JSON.stringify(cacheItem) !== JSON.stringify(dictionaryItem)) {
            setInput(dictionaryItem)
            setCacheItem(dictionaryItem)
        } else {
            setCacheItem(dictionaryItem)
        }
    })

    const submitDictionaryItem = (e: FormEvent) => {
        e.preventDefault()
        store.clearValidation()
        runLocalInputValidations()
        if (store.submitDictionaryItem(inputs, cacheItem)) {
            clearLocalInputValidations()
            clearLocalInputFields()
        }
    }

    const removeDictionaryItem = () => {
        if (isFirstRow) {
            clearLocalInputFields()
        } else {
            store.removeDictionaryItem(inputs)
        }
    }

    const checkLocalInputToForDuplicates = (
        e: React.ChangeEvent<HTMLInputElement>
    ) => {
        if (store.availableDictionaries[e.target.value]) {
            e.target.setCustomValidity(
                `Duplicate 'To' fields are not allowed: value '${
                    e.target.value
                }' already exists`
            )
        } else {
            e.target.setCustomValidity('')
        }
    }

    const handleLocalInputToChange = (
        e: React.ChangeEvent<HTMLInputElement>
    ) => {
        store.clearValidation()
        checkLocalInputToForDuplicates(e)
        setInput({
            from: e.target.value,
            to: inputs.to
        })
    }

    const localInputToIsDuplicate = !!store.validationErrors[
        Validation.KEY_ALREADY_EXISTS
    ][inputs.from]

    return (
        <tr
            className={classnames(styles.row, {
                [styles.editable]: editable
            })}
        >
            <td>
                <form onSubmit={submitDictionaryItem}>
                    <input
                        className={classnames({
                            [styles.inputValidationError]:
                                editable && localInputToIsDuplicate
                        })}
                        disabled={!editable}
                        required
                        ref={fromInput}
                        onBlur={resetLocalInputFrom}
                        type="text"
                        value={inputs.from}
                        onChange={handleLocalInputToChange}
                    />
                </form>
            </td>
            <td>
                <form onSubmit={submitDictionaryItem}>
                    <input
                        disabled={!editable}
                        required
                        ref={toInput}
                        onBlur={resetLocalInputTo}
                        type="text"
                        value={inputs.to}
                        onChange={e =>
                            setInput({
                                to: e.target.value,
                                from: inputs.from
                            })
                        }
                    />
                </form>
            </td>
            {editable && (
                <td
                    className={classnames(styles.controls, {
                        [styles.editable]: editable
                    })}
                >
                    <img
                        className={styles.controlButton}
                        src={checkmark}
                        alt="Submit"
                        onMouseDown={submitDictionaryItem}
                    />
                    <img
                        className={styles.controlButton}
                        src={cross}
                        alt="Delete"
                        onClick={removeDictionaryItem}
                    />
                </td>
            )}
        </tr>
    )
}

export default inject('store')(observer(TableRow)) as React.FunctionComponent<
    Omit<Props, 'store'>
>
