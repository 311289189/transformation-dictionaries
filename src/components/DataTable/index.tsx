import * as React from 'react'
import styles from './styles.css'
import classnames from 'classnames'

import { IStore } from '../../store'
import { observer } from 'mobx-react-lite'
import { inject } from 'mobx-react'
import TableRow from './TableRow'

interface Props {
    store: IStore
    editable?: boolean
}

export interface DictionaryMapping {
    from: string
    to: string
}

const DataTable = ({ store, editable = false }: Props) => {
    const dictionaries: DictionaryMapping[] = Object.keys(
        store.availableDictionaries
    ).map(key => ({
        from: key,
        to: store.availableDictionaries[key]
    }))

    if (editable) {
        dictionaries.unshift({ from: '', to: '' })
    }

    return (
        <table className={classnames(styles.dataTable)}>
            <tbody>
                {dictionaries.map((item, i) => (
                    <TableRow
                        dictionaryItem={item}
                        editable={editable}
                        key={i}
                        index={i}
                    />
                ))}
            </tbody>
        </table>
    )
}

export default inject('store')(observer(DataTable)) as React.FunctionComponent<
    Omit<Props, 'store'>
>
