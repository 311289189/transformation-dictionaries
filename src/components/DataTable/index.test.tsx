import * as React from 'react'
import DataTable from './index'
import { render } from 'enzyme'
import { Provider } from 'mobx-react'

import StoreClass from '../../store'
import { Store } from '../../store'
import { dictionaryItemFixture } from '../../../test/test/fixtures'

describe('DataTable Component', () => {
    describe('In edit mode', () => {
        it('should match snapshot', () => {
            const wrapper = render(
                <Provider store={StoreClass}>
                    <DataTable editable />
                </Provider>
            )
            expect(wrapper).toMatchSnapshot()
        })
    })

    describe('In edit mode, with items', () => {
        it('should match snapshot', () => {
            const StoreWithItems = new Store()
            StoreWithItems.addDictionaryItem(dictionaryItemFixture)

            const wrapper = render(
                <Provider store={StoreWithItems}>
                    <DataTable editable />
                </Provider>
            )
            expect(wrapper).toMatchSnapshot()
        })
    })

    describe('In read-only mode', () => {
        it('should match snapshot', () => {
            const wrapper = render(
                <Provider store={StoreClass}>
                    <DataTable />
                </Provider>
            )
            expect(wrapper).toMatchSnapshot()
        })
    })
})
