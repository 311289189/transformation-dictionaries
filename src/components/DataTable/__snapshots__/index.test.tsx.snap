// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable Component In edit mode should match snapshot 1`] = `
<table
  class="dataTable"
>
  <tbody>
    <tr
      class="row editable"
    >
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value=""
          />
        </form>
      </td>
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value=""
          />
        </form>
      </td>
      <td
        class="controls editable"
      >
        <img
          alt="Submit"
          class="controlButton"
          src="[object Object]"
        />
        <img
          alt="Delete"
          class="controlButton"
          src="[object Object]"
        />
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`DataTable Component In edit mode, with items should match snapshot 1`] = `
<table
  class="dataTable"
>
  <tbody>
    <tr
      class="row editable"
    >
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value=""
          />
        </form>
      </td>
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value=""
          />
        </form>
      </td>
      <td
        class="controls editable"
      >
        <img
          alt="Submit"
          class="controlButton"
          src="[object Object]"
        />
        <img
          alt="Delete"
          class="controlButton"
          src="[object Object]"
        />
      </td>
    </tr>
    <tr
      class="row editable"
    >
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value="test"
          />
        </form>
      </td>
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value="fixture"
          />
        </form>
      </td>
      <td
        class="controls editable"
      >
        <img
          alt="Submit"
          class="controlButton"
          src="[object Object]"
        />
        <img
          alt="Delete"
          class="controlButton"
          src="[object Object]"
        />
      </td>
    </tr>
    <tr
      class="row editable"
    >
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value="other"
          />
        </form>
      </td>
      <td>
        <form>
          <input
            class="inputCell"
            required=""
            type="text"
            value="fixture"
          />
        </form>
      </td>
      <td
        class="controls editable"
      >
        <img
          alt="Submit"
          class="controlButton"
          src="[object Object]"
        />
        <img
          alt="Delete"
          class="controlButton"
          src="[object Object]"
        />
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`DataTable Component In read-only mode should match snapshot 1`] = `
<table
  class="dataTable"
>
  <tbody>
    <tr
      class="explainerText"
    >
      <td>
        <div>
          Nothing to display.
        </div>
        <br />
        <div>
          Add dictionary items using the cells above, all cells are WSYWYG and can be updated individually, use the green button or hit Enter to submit an entry. Use the red button to delete rows.
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
